### ISSUE TITLE
Database Auth

### ISSUE THREAD

**Issue by kfox1111** [spiffe/spire#4968]
TL;DR: It should be possible to configure a SPIRE Server, when using an upstream SPIRE to fetch a SPIFFE Identity from Upstream to use to secure a connection to its MySQL/PostgreSQL.

When setting up a self hosted HA SPIRE setup, one of the bottom turtles is the MySQL/PostgreSQL database. Communications need to be secured between that database and SPIRE but currently relies on passwords or an external CA for certificates. The same SPIRE server can't reliably give itself credentials. But, SPIRE can come to the rescue still. A low availability root SPIRE instance using SQLite can be deployed that doesn't need to scale very far or have high availability to bootstrap the HA SPIRE instances under it. This root server can also be used to give both the database an identity and the child SPIRE server instances identities so they can establish mutual trust with each other.

**Comment by amartinezfayo** [comment]
Hi @kfox1111, thank you for opening this issue, this is an interesting problem to solve, and I'm sure that there is room for improvement in the current experience for this.
>A low availability root SPIRE instance using SQLite can be deployed that doesn't need to scale very far or have high availability to bootstrap the HA SPIRE instances under it

I think that this could be a good idea, although I'm not quite sure if I'm correctly understanding that you are proposing that this SPIRE instance would be used only for bootstrapping? What would happen when the database connection needs to be re-authenticated?

Involving spiffe-helper here could make things easier. But I would like to understand better that bit from your proposal first.

**Comment by kfox1111** [comment]
The idea would be, anytime the HA SPIRE instance needs a cert to talk to its backend DB, it could fetch an up to date cert from the upstream SPIRE. It would be needed for more then just bootstrapping. spiffe-heleper could be used as part of the implementation, yeah. But would be easier if SPIRE itself had built in support for it.

**Comment by amartinezfayo** [comment]
>The idea would be, anytime the HA SPIRE instance needs a cert to talk to its backend DB, it could fetch an up to date cert from the upstream SPIRE. It would be needed for more then just bootstrapping

I see. So this introduces a dependency on that SPIRE server. If it goes down, the server that depends on it to connect to its database will also be affected.
I think that it would be helpful to have some more clarity on how the workflow for this solution would look like. We should be able to answer questions about 1) how a new server is onboarded? 2) how a server gets a new cert to be able to re-authenticate to the database 3) how is this configured? 4) what are the different failure modes?

**Comment by amartinezfayo** [comment]
SPIRE just added support to AWS IAM-based authentication with AWS RDS backed databases, that may solve some of the challenges here in terms of authenticating to the database. @kfox1111 could this help?

**Comment by kfox1111** [comment]
Only if your in AWS and using AWS RDS. :/

**Comment by kfox1111** [comment]
Yes, it does add a dependency on the upstream server. But you have that to some extent if you have an upstream server anyway. Some of the risks of upstream server being down can be mitigated by setting the TTL on the certificate to be long enough to handle reasonable downtimes of the upstream, same as with the upstream intermediate CA.

**Comment by kfox1111** [comment]
We have an example of using a bitnami mysql server deployed with helm, secured with a spiffe identifier obtained via the spiffe helper here: https://github.com/spiffe/helm-charts-hardened/tree/main/examples/mysql-using-spire. That covers the db side of it. Would just need to get the spire-server using a similar, client certificate to give to the database, and refresh as needed.

**Comment by amartinezfayo** [comment]
Thank you @kfox1111 for the details.
I've been thinking how all these pieces would play together, and what would be the scope of the changes needed in SPIRE to support this scenario natively.
To that end, I've tried to represent this in a diagram:
<img width="416" alt="Datastore-X509Auth-WorkloadAPI" src="https://github.com/spiffe/spire/assets/26970586/96a5bc1b-190d-4797-ab12-5bda2fd46a01">
We can have a MySQL / PostgreSQL database supporting secure connections using X509-SVIDs. Rotation of those certificates could be handled by a SPIFFE-Helper connected to a SPIRE Agent that is connected to a SPIRE Server using a SQLite datastore. We can call this deployment the _db-cert-provider_.
In order to have HA instances of SPIRE Server using that MySQL / PostgreSQL database as the datastore, we would need to  configure the datastore to use X509 authentication and have the X509-SVIDs needed fetched from a WorkloadAPI serving X509-SVIDs from the _db-cert-provider_ deployment. This is the part that is missing today.

In summary, could you confirm that providing support in the "sql" Datastore plugin to be able to configure a MySQL / PostgreSQL datastore using X509-SVIDs fetched from the WorkloadAPI to authenticate to the database is the missing part that would fulfill your requirements?

**Comment by kfox1111** [comment]
Yes. That diagram and description match what I was thinking. Thanks for all of that. It looks great. :)

**Comment by amartinezfayo** [comment]
We have discussed this in our last maintainers sync. Although it would be good to provide a better experience for this scenario, the cost of added complexity and dependency problems in this solution don't seem to be a desired trade-off for the project.
Using SPIFFE-Helper to fetch X509-SVIDs when needed without making changes in SPIRE seems to be a solution that could be applied if there would be a desire of instrumenting a solution like the one described.
Since I don't really see a path forward to support this natively in SPIRE at this point in time without the trade-offs mentioned, I'm closing this issue for now.
Thank you @kfox1111 for bringing this up. Please re-open if needed.

**Comment by kfox1111** [comment]
If spire is handed a cert to auth with the database and if the file changes, spire will reread the file? If so, then I think this issue can be closed. If not, I think there is still work to do even if the solution is to use spiffe-helper?

**Comment by amartinezfayo** [comment]
That's a good point. From what I see in the case of MySQL, the [current code doesn't seem to handle cert updates](https://github.com/spiffe/spire/blob/main/pkg/server/datastore/sqlstore/mysql.go#L137), because it's statically set when the plugin is configured. I'm not sure if it would work in the case of PostgreSQL.
If it's confirmed that client certificate rotation doesn't work for either MySQL or PostgreSQL, maybe that could be handled in a separate issue?

**Comment by amartinezfayo** [comment]
I've filed #5113 to consider the support of certificate rotation in the datastore authentication. I'm closing this.
Thanks again @kfox1111 for bringing this up.
